service/rpc/server.go

```diff
@@ -60,16 +67,21 @@ func (s *ServerImpl) Run() error {
 		return err
 	}
 
+	rpcs := grpc.NewServer()
+	rpcs.Register(s.s)
+
 	go func() {
-		c, err := s.s.listener.Accept()
-		if err != nil {
-			panic(err)
-		}
 		defer s.s.listener.Close()
-
-		rpcs := grpc.NewServer()
-		rpcs.Register(s.s)
-		rpcs.ServeCodec(jsonrpc.NewServerCodec(c))
+		for {
+			c, err := s.s.listener.Accept()
+			if err != nil {
+				panic(err)
```

Comment:
We should no longer simply panic here. Since the listener could be closed elsewhere, if this goroutine is allowed to execute long enough to panic users will get confused. We need to either detect [this error](https://golang.org/src/net/net.go#L373) within `OpError` -- or come up with a scheme where this goroutine is the one that closes the listener to avoid this issue.

Since that error var isn't exported I feel we would have to resort to string matching which seems kind of brittle if that error message ever changes significantly (which is not likely, but still possible I suppose).

The other route would be, instead of closing the listener elsewhere, we signal this goroutine to close it like so:

```go
go func() {
	conns := make([]net.Conn)
	defer func() {
		for _, conn := range conns {
			conn.Close()
		}
		s.s.listener.Close()
	}()
	for {
		c, err := s.s.listener.Accept()
		if err != nil {
			select {
			case <-s.s.stopChan:
				// We were supposed to exit, do nothing and return
				return
			default:
				panic(err)
			}
		}
		conns = append(conns, con)
		go rpcs.ServeCodec(jsonrpc.NewServerCodec(c))
		if !s.s.config.AcceptMulti {
			break
		}
	}
}()
```

And then `Stop` becomes:

```go
func (s *ServerImpl) Stop(kill bool) error {
	err := s.s.debugger.Detach(kill)
	if err != nil {
		return err
	}
	if s.s.config.AcceptMulti {
		close(s.s.stopChan)
	}
	return nil
}
```

Thoughts?